php/TypeScript/PhpLexerBase.ts

```diff
@@ -0,0 +1,161 @@
+import {CommonToken, Lexer, Token, CharStream} from "antlr4";
+// change the imports below accordingly
+import PhpParser from "./PhpParser";
+import PhpLexer from "./PhpLexer";
+
+export default abstract class PhpLexerBase extends Lexer {
+    private AspTags: boolean;
+    protected _scriptTag: boolean;
+    protected _styleTag: boolean;
+    private _heredocIdentifier: string | undefined;
+    private _prevTokenType: number;
+    private _htmlNameText: string | undefined;
+    private _phpScript: boolean;
+    private _insideString: boolean;
+    protected _mode: number
+    protected _channel: number
+
+    static DEFAULT_MODE = 0;
+    protected static MORE = -2;
+    protected static SKIP = -3;
+
+    protected static DEFAULT_TOKEN_CHANNEL: number = (Token as any).DEFAULT_CHANNEL;
+    protected static HIDDEN: number = (Token as any).HIDDEN_CHANNEL;
+    protected static MIN_CHAR_VALUE = 0x0000;
+    protected static MAX_CHAR_VALUE = 0x10FFFF;
+
+    constructor(input: CharStream) {
+        super(input);
+        this.AspTags = true;
+        this._scriptTag = false;
+        this._styleTag = false;
+        this._heredocIdentifier = undefined;
+        this._prevTokenType = 0;
+        this._htmlNameText = undefined;
+        this._phpScript = false;
+        this._insideString = false;
+    }
+
+    nextToken() {
+        let token = super.nextToken()
+
+        if (token.type === PhpParser.PHPEnd || token.type === PhpLexer.PHPEndSingleLineComment) {
+            if (this._mode === PhpLexer.SingleLineCommentMode) {
+                // SingleLineCommentMode for such allowed syntax:
+                // // <?php echo "Hello world"; // comment ?>
+                this.popMode();
+            }
+            this.popMode();
+
+            if (token.text === "</script>") {
+                this._phpScript = false;
+                token.type = PhpLexer.HtmlScriptClose;
+            } else {
+                // Add semicolon to the end of statement if it is absent.
+                // For example: <?php echo "Hello world" ?>
+                if (this._prevTokenType === PhpLexer.SemiColon || this._prevTokenType === PhpLexer.Colon || this._prevTokenType === PhpLexer.OpenCurlyBracket || this._prevTokenType === PhpLexer.CloseCurlyBracket) {
+                    token = super.nextToken()
+                } else {
+                    token = new CommonToken(undefined, PhpLexer.SemiColon, undefined, undefined, undefined);
+                    token.text = ';';
```

Comment:
Change these two lines to correspond to that in the CSharp port: `token.type = PhpLexer.SemiColon;`